File: bin_search.java
// Time Complexity : O(Log(N)) -> Complexity for binary search
// Space Complexity : O (1) -> No extra space have been used
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
//I knew the binary search logic but couldn't implement by myself

// Your code here along with comments explaining your approach

class Solution {
    public int search(int[] nums, int target) {
        if(nums == null || nums.length == 0)
            return -1;
        
        int start = 0;
        int end = nums.length -1;
        
        while(start<=end){
            int mid = start + (end - start)/2;
            
            if(nums[mid] == target) return mid;
            if(nums[start]<=nums[mid]){ // left search
               if(nums[start] <= target && target <=nums[mid]){
                   end = mid -1;
               }else{
                   start = mid + 1;
               } 
            }else{ // right search
                if(nums[mid] <= target && target <= nums[end]){
                    start = mid + 1;
                }else{
                    end = mid -1;
                }
            }
        }
        return -1;
    }
}

File: BinSearch.py
# Time Complexity : O(log(mn))
# Space Complexity : O(1), we only use 5 extra variables. So it's constant time irrespective of the input size
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No.

def searchMatrix(matrix: list[list[int]], target: int) -> bool:
    m = len(matrix)
    n = len(matrix[0])
    l,h = 0, m*n-1
    while l<=h:
        mid = l + (h-l)//2   # Prevent overflows
        r = mid//n;
        c = mid%n;
        if matrix[r][c] == target:
            return True
        elif matrix[r][c] > target:
            h = mid-1
        else:
            l = mid+1